<a href="https://colab.research.google.com/github/s1270144/ShutoH/blob/master/Project8_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**CD Algorithm**

In [2]:
!bash


bash: cannot set terminal process group (59): Inappropriate ioctl for device
bash: no job control in this shell
/content# ls
drive  sample_data
/content# cd drive/MyDrive?project8
bash: cd: drive/MyDrive?project8: No such file or directory
/content# cd drive/MyDrive/Project8
/content/drive/MyDrive/Project8# python runrbm_on_mnist.py
python3: can't open file 'runrbm_on_mnist.py': [Errno 2] No such file or directory
/content/drive/MyDrive/Project8# python run_rbm_on_mnist.py
Iteration 0: error is 91105.3532900263
Iteration 5000: error is 32502.867434379645
Iteration 10000: error is 15311.120417440005
Iteration 15000: error is 12906.984738016212
Training multlogistic regression...
Predicting results...
0
0
0
0
0
0
0
0
0
0
# data: 10000, accuracy: 0.8823
110.29478001594543 s
/content/drive/MyDrive/Project8# 
/content/drive/MyDrive/Project8# 
/content/drive/MyDrive/Project8# 
/content/drive/MyDrive/Project8# ^C


**normal RBM**

In [ ]:
!bash

bash: cannot set terminal process group (59): Inappropriate ioctl for device
bash: no job control in this shell
/content# ls
drive  sample_data
/content# cd drive/MyDrive/Project8
/content/drive/MyDrive/Project8# python run_rbm_on_mnist.py
Iteration 0: error is 91105.3532900263
Iteration 50: error is 33127.58264884032
Iteration 100: error is 30121.969369156184
Iteration 150: error is 29049.49733652214
Iteration 200: error is 27807.65862022603
Iteration 250: error is 26890.517302639335
Iteration 300: error is 26061.86297647029
Iteration 350: error is 24386.275285236337
Iteration 400: error is 23260.68565515294
Iteration 450: error is 21865.796774608993
Iteration 500: error is 20621.498658306813
Iteration 550: error is 19765.269379431727
Iteration 600: error is 19459.878283208378
Iteration 650: error is 18919.654406083573
Iteration 700: error is 18298.793506731836
Iteration 750: error is 18318.319373294424
Iteration 800: error is 17559.509544132172
Iteration 850: error is 17298.282951961

CD algorithm

In [ ]:
# CD
    def learn(self, dataIn, rate=0.01, maxIter=3000, wDecay=0.001, verbose=True, useAllData=False):

        data = np.copy(dataIn).astype(np.float64)
        data = np.insert(data, 0, 1, axis=1)
        #data = self.normData(data)
        numData = data.shape[0]

        for it in range(maxIter):

            #np.random.seed(0)

            if useAllData:
                subData = data
            else:
                ind = np.random.uniform(size=self.batchSize, high=data.shape[0]).astype(int)
                subData = data[ind,:]

            if it == 0 or it % 100 == 0:

                forwdHiddenIn = np.dot(subData, self.weights)
                forwdHiddenProb = self.actFunc(forwdHiddenIn)
                forwdHiddenState = (forwdHiddenProb > np.random.rand(self.batchSize, self.numH+1)).astype(int)

                brac0 = np.dot(subData.T, forwdHiddenProb)

                bckwdVisibleIn = np.dot(forwdHiddenState, self.weights.T)
                bckwdVisibleProb = self.actFunc(bckwdVisibleIn)
                bckwdVisibleProb[:,0] = 1
                bckwdHiddenAct = np.dot(bckwdVisibleProb, self.weights)
                bckwdHiddenProb = self.actFunc(bckwdHiddenAct)

                brac1 = np.dot(bckwdVisibleProb.T, bckwdHiddenProb)


            #print("weights:", self.weights)
            #print("delta(weights):", self.rate * ((brac0 - brac1) / self.batchSize))

            #print("batchSize: ", self.batchSize)

            self.weights += rate * (brac0 - brac1) / self.batchSize
            if wDecay != 0:
                self.weights -= rate * wDecay * self.weights

            if verbose and it % 5000 == 0:
                error = np.sum((subData - bckwdVisibleProb)**2)
                print("Iteration %s: error is %s" % (it, error))


By reducing the number of samplings to once every few times, the execution time was considerably shorter than that of normal RBM, and the accuracy was also considerably higher.

CD Algorithm:
*   accuracy: 0.8823
*   Time: 110.29478001594543 s

normal RBM:
*   accuracy: 0.8988
*   Time: 1054.5390467643738 s

Reference: https://github.com/stevejosborne/rbm-examples